### PR TITLE
add ptb model example

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

support ptb language model

## How was this patch tested?

Test the convergence of the model, and compare with tensorflow

## Related links or issues (optional)
https://www.tensorflow.org/tutorials/recurrent#language_modeling

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/LabeledSentenceToSample.scala [hunk 1]
```diff
@@ -90,31 +110,40 @@ class LabeledSentenceToSample[T: ClassTag](vocabLength: Int,
        */
 
 
-      val startTokenIndex = sentence.getData(0)
-      val endTokenIndex = if (labelLength == 1) 0
+        val startTokenIndex = sentence.getData(0)
+        val endTokenIndex = if (labelLength == 1) 0
         else ev.toType[Int](sentence.getLabel(sentence.labelLength - 1))
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/LabeledSentenceToSample.scala]: ident

**Comment by jason-dai** [comment]
change the example name to languagemodel, and add a readme

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/LabeledSentenceToSample.scala (hunk 1 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/LabeledSentenceToSample.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBWordLM.scala [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        param.dataFolder, param.vocabSize)
+
+      val trainSet = DataSet.rdd(sc.parallelize(
+        SequencePreprocess.reader(trainData, param.numSteps)))
+          .transform(TextToLabeledSentence[Float](param.numSteps))
+          .transform(LabeledSentenceToSample[Float](
+            oneHot = false,
+            fixDataLength = None,
+            fixLabelLength = None))
+        .transform(SampleToMiniBatch[Float](param.batchSize))
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBWordLM.scala]: Please don't use DataSet and SampleToMiniBatch, just create a RDD[Sample]

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBWordLM.scala [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+      val validationSet = DataSet.rdd(sc.parallelize(
+        SequencePreprocess.reader(validData, param.numSteps)))
+        .transform(TextToLabeledSentence[Float](param.numSteps))
+        .transform(LabeledSentenceToSample[Float](
+          oneHot = false,
+          fixDataLength = None,
+          fixLabelLength = None))
+        .transform(SampleToMiniBatch[Float](param.batchSize))
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBWordLM.scala]: Please don't use DataSet and SampleToMiniBatch, just create a RDD[Sample]

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBWordLM.scala (hunk 2 above)
Comment (qiuxin2012) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBWordLM.scala]: Add cache the RDD[Sample]

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/LabeledSentenceToSample.scala [hunk 4]
```diff
@@ -90,31 +110,40 @@ class LabeledSentenceToSample[T: ClassTag](vocabLength: Int,
...
+        }
+        while (i < labelLength) {
+          labelBuffer(i) = ev.plus(startTokenIndex, ev.fromType[Float](1.0f))
+          i += 1
+        }
+      } else {
+        feature.resize(dataLength).zero
+        label.resize(labelLength).zero
+
+        val featureBuffer = feature.storage().array()
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/LabeledSentenceToSample.scala]: Just create a new feature and label

**Comment by zhangxiaoli73** [comment]
jenkins pass

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBWordLM.scala (hunk 2 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBWordLM.scala]: this method will increase memory cost.

**Pull request merged**